### ISSUE TITLE
some issues between gradient_accumulation_steps and loss

### ISSUE THREAD

**Issue by Chen-GX** [Qihoo360/360-LLaMA-Factory#1]
### Reminder

- [X] I have read the README and searched the existing issues.

### System Info

accelerate                        1.0.1
aiohappyeyeballs                  2.4.4
aiohttp                           3.11.11
aiosignal                         1.3.2
annotated-types                   0.7.0
anthropic                         0.42.0
anyio                             4.7.0
asttokens                         3.0.0
attrs                             24.3.0
bitsandbytes                      0.45.0
certifi                           2024.12.14
charset-normalizer                3.4.1
click                             8.1.8
cloudpickle                       3.1.0
compressed-tensors                0.8.0
contourpy                         1.3.1
cuda-python                       12.6.2.post1
cycler                            0.12.1
datasets                          3.1.0
decorator                         5.1.1
decord                            0.6.0
deepspeed                         0.15.0
dill                              0.3.8
diskcache                         5.6.3
distro                            1.9.0
docstring_parser                  0.16
einops                            0.8.0
executing                         2.1.0
fastapi                           0.115.6
filelock                          3.16.1
flash-attn                        2.7.2.post1
flashinfer                        0.1.6+cu124torch2.4
fonttools                         4.55.3
frozenlist                        1.5.0
fsspec                            2024.9.0
gguf                              0.10.0
h11                               0.14.0
hf_transfer                       0.1.8
hjson                             3.1.0
httpcore                          1.0.7
httptools                         0.6.4
httpx                             0.27.2
huggingface-hub                   0.27.0
idna                              3.10
importlib_metadata                8.5.0
iniconfig                         2.0.0
interegular                       0.3.3
ipython                           8.31.0
jedi                              0.19.2
Jinja2                            3.1.5
jiter                             0.8.2
jsonschema                        4.23.0
jsonschema-specifications         2024.10.1
kiwisolver                        1.4.8
lark                              1.2.2
litellm                           1.55.12
llvmlite                          0.43.0
lm-format-enforcer                0.10.9
markdown-it-py                    3.0.0
MarkupSafe                        3.0.2
matplotlib                        3.10.0
matplotlib-inline                 0.1.7
mdurl                             0.1.2
mistral_common                    1.5.1
modelscope                        1.21.0
mpmath                            1.3.0
msgpack                           1.1.0
msgspec                           0.18.6
multidict                         6.1.0
multiprocess                      0.70.16
nest-asyncio                      1.6.0
networkx                          3.4.2
ninja                             1.11.1.3
numba                             0.60.0
numpy                             1.26.4
nvidia-cublas-cu12                12.4.5.8
nvidia-cuda-cupti-cu12            12.4.127
nvidia-cuda-nvrtc-cu12            12.4.127
nvidia-cuda-runtime-cu12          12.4.127
nvidia-cudnn-cu12                 9.1.0.70
nvidia-cufft-cu12                 11.2.1.3
nvidia-curand-cu12                10.3.5.147
nvidia-cusolver-cu12              11.6.1.9
nvidia-cusparse-cu12              12.3.1.170
nvidia-ml-py                      12.560.30
nvidia-nccl-cu12                  2.21.5
nvidia-nvjitlink-cu12             12.4.127
nvidia-nvtx-cu12                  12.4.127
openai                            1.58.1
opencv-python-headless            4.10.0.84
orjson                            3.10.12
outlines                          0.0.46
packaging                         24.2
pandas                            2.2.3
parso                             0.8.4
partial-json-parser               0.2.1.1.post4
peft                              0.12.0
pexpect                           4.9.0
pillow                            10.4.0
pip                               24.2
pluggy                            1.5.0
prometheus_client                 0.21.1
prometheus-fastapi-instrumentator 7.0.0
prompt_toolkit                    3.0.48
propcache                         0.2.1
protobuf                          5.29.2
psutil                            6.1.1
ptyprocess                        0.7.0
pure_eval                         0.2.3
py-cpuinfo                        9.0.0
pyairports                        2.1.1
pyarrow                           18.1.0
pybind11                          2.13.6
pycountry                         24.6.1
pydantic                          2.10.4
pydantic_core                     2.27.2
Pygments                          2.18.0
pyparsing                         3.2.1
pytest                            8.3.4
python-dateutil                   2.9.0.post0
python-dotenv                     1.0.1
python-multipart                  0.0.20
pytz                              2024.2
PyYAML                            6.0.2
pyzmq                             26.2.0
ray                               2.40.0
referencing                       0.35.1
regex                             2024.11.6
requests                          2.32.3
rich                              13.9.4
ring-flash-attn                   0.1.3
rpds-py                           0.22.3
safetensors                       0.4.5
sentencepiece                     0.2.0
setproctitle                      1.3.4
setuptools                        75.1.0
sgl-kernel                        0.0.2.post9
sglang                            0.4.1
shtab                             1.7.1
six                               1.17.0
sniffio                           1.3.1
stack-data                        0.6.3
starlette                         0.41.3
sympy                             1.13.1
tiktoken                          0.7.0
tokenizers                        0.20.3
torch                             2.5.1
torchao                           0.7.0
torchvision                       0.20.1
tqdm                              4.67.1
traitlets                         5.14.3
transformers                      4.46.1
triton                            3.1.0
trl                               0.9.6
typeguard                         4.4.1
typing_extensions                 4.12.2
tyro                              0.9.5
tzdata                            2024.2
urllib3                           2.3.0
uvicorn                           0.34.0
uvloop                            0.21.0
vllm                              0.6.4.post1
watchfiles                        1.0.3
wcwidth                           0.2.13
websockets                        14.1
wheel                             0.44.0
xformers                          0.0.28.post3
xgrammar                          0.1.8
xxhash                            3.5.0
yarl                              1.18.3
zipp                              3.21.0

### Reproduction

```finetuning_type=full
learning_rate=5e-6

total_bsz=2
sequence_parallel_size=4
per_device_train_batch_size=1

${deepspeed_env} --num_gpus ${GPU_NUM} ../360-LLaMA-Factory/src/train.py \
    --deepspeed ${deepspeed_config_file} \
    --stage sft \
    --do_train \
    --flash_attn fa2 \
    --packing True \
    --sequence_parallel_size ${sequence_parallel_size} \
    --model_name_or_path ${model_name_or_path} \
    --dataset_dir ${dataset_dir}\
    --dataset ${dataset} \
    --template qwen \
    --finetuning_type ${finetuning_type} \
    --save_safetensors \
    --output_dir ${output_dir} \
    --overwrite_cache \
    --max_new_tokens 32768 \
    --cutoff_len 32768 \
    --per_device_train_batch_size ${per_device_train_batch_size} \
    --gradient_accumulation_steps ${gradient_accumulation_steps} \
    --warmup_ratio 0.03 \
    --weight_decay 0. \
    --lr_scheduler_type cosine \
    --logging_steps 10 \
    --save_steps 20 \
    --learning_rate ${learning_rate} \
    --num_train_epochs 3.0 \
    --dataloader_num_workers 16 \
    --preprocessing_num_workers 128 \
    --ddp_timeout 180000000 \
    --seed $seed \
    --plot_loss \
    --gradient_checkpointing \
    --save_only_model \
    --bf16
```

### Expected behavior

Abnormally large loss when gradient_accumulation_steps is greater than 1.

For example, when gradient_accumulation_steps=16
```
{"current_steps": 5, "total_steps": 198, "loss": 14.8746, "lr": 4.166666666666667e-06, "epoch": 0.07568590350047304, "percentage": 2.53, "elapsed_time": "0:04:56", "remaining_time": "3:10:34"}
{"current_steps": 10, "total_steps": 198, "loss": 14.2352, "lr": 4.994647308096509e-06, "epoch": 0.15137180700094607, "percentage": 5.05, "elapsed_time": "0:09:44", "remaining_time": "3:03:05"}
{"current_steps": 15, "total_steps": 198, "loss": 14.5774, "lr": 4.972941274911953e-06, "epoch": 0.22705771050141912, "percentage": 7.58, "elapsed_time": "0:14:55", "remaining_time": "3:02:03"}
{"current_steps": 20, "total_steps": 198, "loss": 14.5221, "lr": 4.9346924481933345e-06, "epoch": 0.30274361400189215, "percentage": 10.1, "elapsed_time": "0:19:43", "remaining_time": "2:55:36"}
```

when gradient_accumulation_steps=1
```
{"current_steps": 10, "total_steps": 1587, "loss": 0.8529, "lr": 1.0416666666666667e-06, "epoch": 0.01890359168241966, "percentage": 0.63, "elapsed_time": "0:05:40", "remaining_time": "14:54:55"}
{"current_steps": 20, "total_steps": 1587, "loss": 0.8395, "lr": 2.0833333333333334e-06, "epoch": 0.03780718336483932, "percentage": 1.26, "elapsed_time": "0:10:54", "remaining_time": "14:14:00"}
```

**Comment by HaoshengZou** [comment]
It seems to be a problem with `transformers 4.46.1`. We have reproduced similar behavior with `transformers 4.46.1`.
We have tested with `transformers < 4.46` and everything is OK.
Could you try `transformers < 4.46` for now while we try to fix it with `transformers >= 4.46`?

**Comment by Chen-GX** [comment]
thanks for your help. i get the expected behavior between gradient_accumulate_steps and loss in `transformers < 4.46`. it seems to be triggerd by `transformers >= 4.46`.

i have an another trivial question about the `total train batch size`.

when i set `gradient_accumulate_steps=16`, `Instantaneous batch size per device = 1`, `GPU_NUM=8`, `sequence_parallel_size=4`, the `total train batch size` looks the same as `gradient_accumulate_steps * GPU_NUM * Instantaneous batch size per device`.  is it a expected result?

i am not very familiar with `sequence parallelism`, `sequence_parallel_size=4` looks like using 4 GPU to deal with a sequence. so there are only 2 logical gpu for training? (8 / 4) 

pardon me if this sounds obvious.

**Comment by HaoshengZou** [comment]
> the total train batch size looks the same as gradient_accumulate_steps * GPU_NUM * Instantaneous batch size per device. is it a expected result?

Yes.

> sequence_parallel_size=4 looks like using 4 GPU to deal with a sequence. so there are only 2 logical gpu for training? (8 / 4)

Logically, Yes. 

The answers are both yes because in our implementation, we split the sequences for SP so that the number of data in the final dataset is already multiplied by SP (you'll notice the change of dataset size in the log). Therefore `total train batch size = gradient_accumulate_steps * GPU_NUM * Instantaneous batch size per device`.

**Comment by Chen-GX** [comment]
oh, i see, thanks for your patience

**Comment by HaoshengZou** [comment]
We've verified that it's the problem with `transformers == 4.46.x`. LLaMA-Factory has already banned `transformers == 4.46.x` (https://github.com/hiyouga/LLaMA-Factory/pull/6628/). We have tested with LLaMA-Factory's supported `transformer` versions and everything is OK.

**Comment by DSTTSD** [comment]
I am using transformers==4.51.3, and i met the same problem during sft training, here is the the loss without sp:
<img width="148" height="112" alt="Image" src="https://github.com/user-attachments/assets/6140d04a-8c36-4600-9494-764a18e96016" />

And here is the loss when SP=4:
<img width="139" height="213" alt="Image" src="https://github.com/user-attachments/assets/ea51a5ca-17f1-4e42-81fa-ce7bcd84ce3c" />
